### PR TITLE
Typed Builder implementation Part 11: `LET-IN` 

### LINKED ISSUE #1955: Implement `TypedBuilder` for `LetInEx`
Also resolves TODOs in `TestApalacheOper`.

### LINKED ISSUE #1956: Rename `BuilderUtil.allUsed` and `BuilderUtil.allBound` with a `get-` prefix
Some methods in `BuilderUtil` are getters of internal state, which means that calling them at different time yields different results, e.g.
```scala
for {
A1 <- allUsed
...
A2 <- allUsed // May be different from A1
} yield ...

```
We should rename them with a `get-` prefix and add more expansive documentation, to make it more intuitive.

### PR DESCRIPTION
**Pull request by Kukovec**
- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [ ] Entries added to [./unreleased/][unreleased] for any new functionality

closes #1955 
closes #1956 

[unreleased]: https://github.com/informalsystems/apalache/tree/unstable/.unreleased

### CODE REVIEW

**Review by thpani — CHANGES_REQUESTED** [review]
I'll echo @shonfeder's previous concerns (https://github.com/informalsystems/apalache/pull/1937#pullrequestreview-1033155210) that this should really be 2 separate, orthogonal PRs.
Can you please cherry-pick those two commits into different branches and open separate PRs?

I'd also appreciate some context in the PR description about what a PR is trying to achieve and how (also useful for future reference, if we have to do archeology).

**Comment by shonfeder** [comment]
I agree with @thpani that it would have be nicer in principle to split this into two separate PRs, and I would support blocking on that PR refactor if this were a large PR. But since it is pretty small anyhow I don't think it's worth the added friction here. 

> I don't really consider it worthwhile to go through all of the overhead of turning one-click refactoring (https://github.com/informalsystems/apalache/issues/1956) into a PR. It also seems like a waste of time for reviewers to have to go through 2 context-switches instead of 1 for reviewing

IMO, the overhead to open a PR should be trivial enough that this isn't much of a concern, and I think we can really improve the quality and velocity of PR reviews by separating automatic changes from the nuanced changes to business logic (and by separating distinct logical changes). Just my 2 cents there.

**Comment by thpani** [comment]
> I agree with @thpani that it would have been nice in principle to split this into two separate PRs, and I would support blocking on that PR refactor if this were a large PR. But since it is pretty small anyhow I don't think it's worth the added friction here.

I'm not concerned at all if small fixes sneak in once in a while, which is why I reviewed #1937 despite the same issue. I'm more concerned about repeated PRs that could've been split, and where reviewing would've been much easier if they had been split.
I shouldn't have blocked this though, if someone else wants to review.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala
Comment (shonfeder) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala:69]: Much clearer, thank you!

Just to confirm, the call to the `variable` instruction below doesn't update the state?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala
Comment (shonfeder) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala:42]: So, iiuc, this acccess to `set` is also changing the state, right? Then wouldn't it be clearer to have this something like:

```suggestion
    setEx <- getNextSet
```

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala
Comment (shonfeder) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala:127]: Very helpful. Thank you!

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala
Comment (shonfeder) [inline tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala:234]: appreciate the message here! I suggest throwing an exception here with the comment as the message instead of returning the type. That will ensure that if this ever does get hit for some reason we'll get the expected error rather than returning an int arbitrarily :)

**Comment by Kukovec** [comment]
> [...] I'm more concerned about repeated PRs [...] where reviewing would've been much easier if they had been split.

That's a fair concern, but I honestly don't think it applies here. The time it takes to review changes to `BuilderUtil` is the time it takes to scroll to the end of the file. This is primarily a PR that implements https://github.com/informalsystems/apalache/issues/1955.

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala
Comment (shonfeder) [inline tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala:435]: This is a very big test, imo, and i cannot  tell by reading this part what properties or behaviors are actually being tested. I know that the test is testing `let-in`, but other than than I can't tell what kind of things are mean to hold here. 

I suggest either refactoring the test into a few distinct property tests that clearly communicate and exemplify what is being tested or adding comments here that explain what is being tested in each chunk. The former is much more preferable, IMO, but I'd be fine with just the latter.

**Review by shonfeder — CHANGES_REQUESTED** [review]
Looking good! I have a few questions, and at least two suggestions I'd like to see addressed before merging.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala:42]: All `TBuilderInstruction`s "change state". The reason the parameters are named the way they are is to keep the intuition that, e.g. `builder.exists(x, S, p)` creates a `OperEx(exists, y, T, q)`-like expression, as not to bleed implementation details into the interface.

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala
Comment (Kukovec) [reply tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala:234]: That's fair, I'll do that.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala:69]: It does, they all do.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala
Comment (shonfeder) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala:42]: This makes sense. So the set/get prefix can be saved just for the utilities that inspect the environment or update it, not for the operators that create the expressions. Thanks for explaining.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala
Comment (shonfeder) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala:187]: I think that’s worth doing here.

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala
Comment (Kukovec) [reply tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala]: I've split the tests and added comments, most notably to the method `runLetTests`.

**Review by shonfeder — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**
